### ISSUE TITLE
Very poor Uniform() sampling near floating 0.0

### ISSUE THREAD

**Issue by obilaniu** [pytorch/pytorch#16706]
## 🐛 Bug

When using uniformly-distributed random number generation, especially using `float32`, over a range including 0, the sampling code ignores the extra precision of floating-point around 0, leading to very bad sampling.

The generated floats also have at most 24 (`float32`) or 53 (`float64`) bits of entropy.

**Consequence 1:** Multiplying any PyTorch float32 `uniform_()` by a large integer like `2**33` is guaranteed to give an integer.
**Consequence 2:** It is actually possible to generate an exact `0.0` using the `float32` `uniform_()`, with probability 2**-24, whereas that should actually have been 2**-149.
**Consequence 3:** It is not possible to generate a `float32` strictly between `0.0` and `2**-24`, whereas it should have been possible with probability 2**-24.

## To Reproduce

*Test 1:* If the following is `!= 0`, the RNG sucks: `torch.FloatTensor(int(1e8)).uniform_().eq(0).sum()`
*Test 2:* If the following is `== -inf`, the RNG sucks: `torch.FloatTensor(int(1e8)).uniform_().log().sum()`

## Expected behavior

*Test 1:* A return of `0`.
*Test 2:* A finite, negative value.

## Environment

_Not relevant._

 - PyTorch Version 1.1.0a0+c865d46, but the same problem dates back to the beginning of the history of this code, which is over a year.

## Additional context

This has serious knock-on effects on many other distributions that use `uniform_()` as a basis. In particular, `log(uniform_())` has a 2**-24 chance of producing `-inf`, and `log2(uniform_())` cannot produce a number `< -24` (except `-inf`).

**Comment by soumith** [comment]
from reading some review comments on https://github.com/pytorch/pytorch/pull/16604 and it's previous versions, I'm just wondering if that PR would fix it, as it effectively rewrites a bunch of stuff to use Philox rather than the original Mersienne-Twister implementation from torch5/7

**Comment by ngimel** [comment]
CPU-side generation unfortunately starts with throwing away 8 bits out of 32, thus resulting in at most 24 bits of entropy, as @obilaniu notes https://github.com/pytorch/pytorch/blob/517c7c98610402e2746586c78987c64c28e024aa/aten/src/TH/THRandom.cpp#L227-L242. @syed-ahmed 's refactor might address that, but I don't remember the details. 
GPU-side generation does not do that, so should be better in this respect 
```
QUALIFIERS float _curand_uniform(unsigned int x)
{
    return x * CURAND_2POW32_INV + (CURAND_2POW32_INV/2.0f);
}
```
As for consequence 2, probability of generating 0 should be 2^-149, I don't know how to achieve it if starting with 32 bits of randomness. There are 2^32 possible outcomes, at least one of them should be mapped to zero, which puts probability of generating 0 to at least 2^-32.

**Comment by ngimel** [comment]
Now that I look at it, not only #16604 does not solve it, it likely breaks GPU in similar way, by discarding random bits https://github.com/pytorch/pytorch/pull/16604/files#diff-630e4621b02f4915fa4813e67c6a61b7R62
cc @syed-ahmed

**Comment by obilaniu** [comment]
I concur with @ngimel ; #16604 has the same problems, and on GPU the code that converts cuRAND's high-quality uniform samples from `(0, 1]` to `[0, 1)` seems to discard all the hard work that cuRAND put into the near-0 uniforms.

cuRAND explicitly documents that it may take a variable amount of entropy (not just 24 bits) from the RNG in order to generate any given float; presumably this is an allusion to near-0 sampling.

**Comment by ngimel** [comment]
There's no hard work in curand, basically all there is is what I posted above
```
return x * CURAND_2POW32_INV + (CURAND_2POW32_INV/2.0f);
```
It is slightly better near 0 because there actually will be numbers generated between 0 and 2^-24, but consequence 1 and consequence 2 would still be present even with curand uniform sampling.

**Comment by obilaniu** [comment]
@ngimel Well, if referencing the [cuRAND documentation](https://docs.nvidia.com/cuda/curand/device-api-overview.html#distributions):

> **3.1.4. Distributions**
>
>     __device__ float 
>     curand_uniform (curandState_t *state)
>
> This function returns a sequence of pseudorandom floats uniformly distributed between 0.0 and 1.0. It may return from 0.0 to 1.0, where 1.0 is included and 0.0 is excluded. Distribution functions may use **any number** of unsigned integer values from a basic generator. **The number of values consumed is not guaranteed to be fixed.**

_(emphasis added)_

This feels like it's clearly alluding to a greater-than-32 (and even variable) number of bits of entropy being drawn from the generator per uniform float, presumably to refine the sampling near `0.0`, but I may be mistaken.

**Comment by ngimel** [comment]
All device-side generation is in the open-source headers. What do you believe - documentation or the code? Don't tell anyone, but curand_uniform_double with curandStateMtgp uses 32 bits of entropy to generate what it calls uniformly distributed double 
```
QUALIFIERS double curand_uniform_double(curandStateMtgp32_t *state)
{
    return _curand_uniform_double(curand(state));
}
```
and, incidentally, mtgp32 is what pytorch uses currently.

**Comment by obilaniu** [comment]
Beurk. Thanks for the heads up.

Well, at least 32 is better than 24.

**Comment by obilaniu** [comment]
After much thinking (with some inspiration from the link @syed-ahmed shared), I believe I have a simple algorithm that produces _high-quality_, _unbiased_ (*) `float32` draws from the uniform distribution `[0.0, 1.0]`, closed at both ends (PyTorch's is open at `1.0`). It would be about 5x slower than a "regular", bad-quality `float32` draw (because I need 5 of those bad-quality numbers), but one can trade between speed and quality. A `float64` version exists as well.

The problem is that it depends on two things not in PyTorch, both quite ugly:

- A temporary change in IEEE Std 754 **rounding mode**.
- **Type-punning** / bit-cast / `reinterpret_cast` [to access a float's representation](https://discuss.pytorch.org/t/reinterpret-pytorch-array-as-a-different-dtype/24256).

Will either ever happen in PyTorch?

_(*) By "high-quality" and "unbiased", I mean that the second-smallest value that my algorithm can return is 2^-120, and the smallest is 0, both with probability 2^-120; Moreover, the resulting bias is <=2^-120, or essentially insignificant._

**Comment by ezyang** [comment]
Reinterpret cast is OK (Python is not strict aliasing safe, and neither is PyTorch), but changing the rounding mode sounds very slow, so that will be a hard sell.

**Comment by obilaniu** [comment]
Alright, I've managed to reformulate my code to avoid rounding mode changes and type-punning.

Attached ([utils.zip](https://github.com/pytorch/pytorch/files/2849866/utils.zip)) are extremely numerically stable routines for Uniform(0,1) and logUniform(0,1) sampling. For `float32`,

- `uniform_ultra_(x)` will return 0 with probability 2^-126 and 1 with probability 2^-25.
- `loguniform_ultra_(x)` will return 0 with probability 2^-127 (because the logarithms of real numbers very near 1 are extremely tiny) and will never return `-inf` because logarithms beyond `FLT_MIN` exponentially more improbable than even those with log2(x) < -126.

Right now these routines are much slower than the normal samplers, but that's mostly because of all those unnecessary large intermediate buffers being created.

@syed-ahmed @ngimel PyTorch & NVIDIA, you may use them as you see fit, but I'd love to receive credit for them if you find them useful :-)

**Comment by ngimel** [comment]
Fun fact: std gets 0-inclusive 1-exclusive wrong http://open-std.org/JTC1/SC22/WG21/docs/lwg-active.html#2524 (hat tip to @syed-ahmed )

**Comment by colesbury** [comment]
I'm not convinced we should change this. This is pretty common behavior for pseudo-random number generators. Most implementations I've seen divide by 2^k to convert a uint64 or uint32 to a double or float, including TF, Julia, NumPy, Go, Boost random, and Marsaglia's KISS. L’Ecuyer's SSJ generates doubles by dividing by 2^32+1 by default.

TensorFlow and NumPy have nearly the same implementation for uniform float and double on [0, 1). [1] Using k=53 for doubles and k=24 for floats has the nice property that every generated number has equal probability. It also means that you can convert a uniform on [0, 1) to (0, 1] by computing 1 - x. This doesn't work with the division by 2^32.  I'll also note that dividing by 2^32 instead of 2^24 for float32 increases the entropy by less than one bit.

If your problem is that you don't want 0 as a value, the fix is to not include 0. L’Ecuyer recommends using a generator that outputs (0, 1) for this reason, but most languages and popular libraries now seem to use [0, 1), with the notable exception of Matlab.

If you want to sample from the exponential distribution, sample directly from the exponential distribution (torch.distributions.Exponential).

[1] TF and NumPy only generate 2^23 unique float32 values instead of 2^24.

![](https://imgs.xkcd.com/comics/random_number.png)

**Comment by obilaniu** [comment]
@colesbury On the contrary, I think this problem bites us regularly, and was about to nominate #22442 as yet another example of bad PRNG possibly connected to this problem.

- That TF, Julia, NumPy, Go, Boost, Marsaglia-KISS and L'Ecuyer-SSJ divide by `2**k` (`+1`?) is their problem. PyTorch doing the same is commendable, so long as it isn't a problem. But it is a problem, and that makes it *our* problem.
- Marsaglia and L'Ecuyer are _extremely authoritative_ for processes generating pseudo-random integers on given ranges, but the naïveté of their conversions to floating-point types betrays a disinterest that possibly stems from an incomplete understanding of IEEE Std 754 floating-point behaviour and how it can corrupt the uniformness of their random numbers.
- > Using k=53 for doubles and k=24 for floats has the nice property that every generated number has equal probability.
  
  But that's not the property we're looking for. Generating `1.00` 50% of the time and `0.50` 50% of the time would also satisfy that property, but we're actually interested in uniform random sampling. But because floating-point numbers are not uniformly spaced (they're more tightly packed around 0), in order to properly sample uniformly from them, one cannot sample uniformly from all floating-point numbers between 0 and 1 inclusive.
- > It also means that you can convert a uniform on [0, 1) to (0, 1] by computing 1 - x
  
  Which is numerically unstable. Numerically-stable code avoids performing `1-x` when `x` may be small, and uses appropriate routines instead, such as `log1p(-x)` or `expm1(-x)`. This is precisely why my proposed stable uniform random sampling code from Feb. 11, 2019 above carefully uses `log_()` and `log1p_()` when each is called for.
- > If your problem is that you don't want 0 as a value, the fix is to not include 0. L’Ecuyer recommends using a generator that outputs (0, 1) for this reason, but most languages and popular libraries now seem to use [0, 1), with the notable exception of Matlab.
  
  They do that because generating a value in the range `[0, 2**k-1]` and dividing by `2**k` naturally results in a semi-open `[0, 1)` range. My Feb. 11, 2019 `uniform_ultra_() `code samples on the closed range `[0, 1]` while my `log_uniform_ultra_()` code samples on the closed range `[-127, 0]`, but the probability of getting an exact 0 from `uniform_ultra_()` or an exact 0 from `log_uniform_ultra_()` is so low (`p=2**-126`) that it can be neglected. `uniform_ultra_()` *can* return an exact `1`, with probability `2**-24` for `float32`, but since you are not meant to be using it in a subtraction `1-x`, this is not a problem.
  
  _Adding_ a small `epsilon` like `2**-126` is not a solution because it doesn't cure the problem that only `2**24` different `float32` values will ever be generated, with a greater variety near 1 and a lower variety near 0. It also biases the output very slightly, and incorrectly implies that extremely unlikely events (like sampling values `<= 2**-126`) happen unduly often (`p=2**-24`, for instance). It's the machine equivalent of human paranoia.
  
- > If you want to sample from the exponential distribution, sample directly from the exponential distribution (torch.distributions.Exponential).
  
  We don't want to sample from the exponential distribution. We want to sample from the uniform distribution. Floating-point numbers are not exponentially-distributed. Their exponents are geometrically distributed, however. One way to sample from the uniform distribution on `[0,1]` is to sample the **mantissa** from the _uniform_ distribution on the closed range `[0.5, 1.0]` and the **exponent** from the _geometric_ distribution with parameter `p=0.5`, but as of Feb. 11, 2019 the function `.geometric_()` on Git master did not work, so I did not use it.

**Comment by colesbury** [comment]
> colesbury On the contrary, I think this problem bites us regularly, and was about to nominate #22442 as yet another example of bad PRNG possibly connected to this problem.

It's connected only in the sense that your proposed fix would introduce this problem to the CPU random number generator. Gumble-softmax uses something like log(-log(uniform())). The NaNs occurred in CUDA the uniform generator return values of 1. (A 0 value is also problematic, but was not
generated by the CUDA generator at the time).

> That TF, Julia, NumPy, Go, Boost, Marsaglia-KISS and L'Ecuyer-SSJ divide by `2**k` (`+1`?) is their problem. PyTorch doing the same is commendable, so long as it isn't a problem. But it is a problem, and that makes it _our_ problem.

Another interpretation is that if the experts and popular mathematical libraries don't consider this a problem, then it might not be an actual problem.

> Marsaglia and L'Ecuyer are _extremely authoritative_ for processes generating pseudo-random integers on given ranges, but the naïveté of their conversions to floating-point types betrays a disinterest that possibly stems from an incomplete understanding of IEEE Std 754 floating-point behaviour and how it can corrupt the uniformness of their random numbers.

This is an interesting take...

> > Using k=53 for doubles and k=24 for floats has the nice property that every generated number has equal probability.
>   
>   But that's not the property we're looking for. Generating `1.00` 50% of the time and `0.50` 50% of the time would also satisfy that property, but we're actually interested in uniform random sampling. But because floating-point numbers are not uniformly spaced (they're more tightly packed around 0), in order to properly sample uniformly from them, one cannot sample uniformly from all floating-point numbers between 0 and 1 inclusive.

Of course, one cares about precision (and many other properties). But it's not clear that one cares about the extra dynamic range near 0 when sampling from U(0, 1).

> > It also means that you can convert a uniform on [0, 1) to (0, 1] by computing 1 - x
>   
>   Which is numerically unstable. Numerically-stable code avoids performing `1-x` when `x` may be small, and uses appropriate routines instead, such as `log1p(-x)` or `expm1(-x)`. This is precisely why my proposed stable uniform random sampling code from Feb. 11, 2019 above carefully uses `log_()` and `log1p_()` when each is called for.

No, it's precise here. In fact, (1-(1-x)) exactly recovers the original [0, 1) distribution when x is generated the PyTorch/TF/NumPy/etc. way.

> > If your problem is that you don't want 0 as a value, the fix is to not include 0. L’Ecuyer recommends using a generator that outputs (0, 1) for this reason, but most languages and popular libraries now seem to use [0, 1), with the notable exception of Matlab.
>   
>   They do that because generating a value in the range `[0, 2**k-1]` and dividing by `2**k` naturally results in a semi-open `[0, 1)` range. My Feb. 11, 2019 `uniform_ultra_() `code samples on the closed range `[0, 1]` while my `log_uniform_ultra_()` code samples on the closed range `[-127, 0]`, but the probability of getting an exact 0 from `uniform_ultra_()` or an exact 0 from `log_uniform_ultra_()` is so low (`p=2**-126`) that it can be neglected.

This is unlikely to be the reason. As I wrote above, it's trivial to transform the [0, 1) distribution into the (0, 1] distribution when using the PyTorch/TF/NumPy/etc. method.

> `uniform_ultra_()` _can_ return an exact `1`, with probability `2**-24` for `float32`, but since you are not meant to be using it in a subtraction `1-x`, this is not a problem.

The referenced Gumbel-softmax issue (#22442) shows how generating 1 can be just as much a problem as generating 0.

>   _Adding_ a small `epsilon` like `2**-126` is not a solution because it doesn't cure the problem that only `2**24` different `float32` values will ever be generated, with a greater variety near 1 and a lower variety near 0. It also biases the output very slightly, and incorrectly implies that extremely unlikely events (like sampling values `<= 2**-126`) happen unduly often (`p=2**-24`, for instance). It's the machine equivalent of human paranoia.

I agree that adding a small epsilon is not a good solution.

> > If you want to sample from the exponential distribution, sample directly from the exponential distribution (torch.distributions.Exponential).
>   
> We don't want to sample from the exponential distribution. We want to sample from the uniform distribution.

Your original post includes "log(uniform_())" as an example. That is sampling from the (negative) exponential distribution.

**Comment by syed-ahmed** [comment]
For the reference, this article gives a good overview on all the various methods that are popular for going from uints to floats: https://experilous.com/1/blog/post/perfect-fast-random-floating-point-numbers

Based on that article and discussion with @ngimel @fritzo @neerajprad , and surveying the other popular libraries like TF, scipy, V8, etc. I suggest the following algorithm (divide and squash the zero):
```
C10_HOST inline dist_acctype<T> operator()(at::CPUGenerator* generator){
    // See Note [Uniform Distribution Algorithm]
    dist_acctype<T> x;
    if(std::is_same<T, double>::value) {
      x = generator->random64() * POW_2_64_INV;
    } else {
      x = generator->random() * POW_2_32_INV;
    }
    if (x == static_cast<T>(1.0)) {
      x = std::nextafter(static_cast<T>(1.0), static_cast<T>(0.0));
    }
    return (x * (b - a) + a);
  }
```

We can discuss more about this algorithm choice, but another problem tied to this is we need to address the test suite which seems to be a little sensitive to the uniform distribution algorithm changes. We have been tracking that here but could open a separate issue for it: https://docs.google.com/document/d/1EqZT6bI2z5TV56h9S8HPcfZJBU64ia9IXTXq7SboO4E/edit.

CC: @fritzo @neerajprad @vishwakftw

**Comment by colesbury** [comment]
@syed-ahmed, I'm confused as to why you're suggesting this algorithm. None of the libraries you reference use that algorithm. They all produce equidistant values (e.g. `x = (generator->random64() >> 11) * POW_2_53_INV)` for doubles). The referenced blog post also suggests producing equidistant values.

**Comment by syed-ahmed** [comment]
Near the end of the article, the author talks about "High Precision Alternative Distribution" method, which is the division method:
>As mentioned near the beginning of the article, all of the above functions sacrifice possible precision that floating point formats are capable of representing for the sake of more perfect uniformity. If you would rather get the higher precision for numbers close to zero, but wish to still avoid division and take advantage of the techniques above, there is a way to do this.

Since we are considering getting more precision near 0.0 in this issue, divide by max uint takes us closer to that without sacrificing much on the performance. And if you do divide by max uint, we need a way to get values in the range [0, 1), and std::nextafter does that (this is also suggested as a solution in the libstd++ [0, 1) problem: [http://open-std.org/JTC1/SC22/WG21/docs/lwg-active.html#2524](http://open-std.org/JTC1/SC22/WG21/docs/lwg-active.html#2524). Although, note that libstd probably uses the 53 bit/23 bit method multiple times through ```std::generate_canonical``` [https://en.cppreference.com/w/cpp/numeric/random/generate_canonical](https://en.cppreference.com/w/cpp/numeric/random/generate_canonical) ).

**Comment by syed-ahmed** [comment]
For the sake of completeness in this issue:  Curand does divide by max uint and uses the `+ 2^32/2.0` trick from http://www.doornik.com/research/randomdouble.pdf and hence has the range (0,1]

```
QUALIFIERS float4 _curand_uniform4(uint4 x)
{
    float4 y;
    y.x = x.x * CURAND_2POW32_INV + (CURAND_2POW32_INV/2.0f);
    y.y = x.y * CURAND_2POW32_INV + (CURAND_2POW32_INV/2.0f);
    y.z = x.z * CURAND_2POW32_INV + (CURAND_2POW32_INV/2.0f);
    y.w = x.w * CURAND_2POW32_INV + (CURAND_2POW32_INV/2.0f);
    return y; 
}
```

**Comment by ngimel** [comment]
But, in a show of wonderful consistency, for doubles it divides by 2^53:
```
QUALIFIERS double _curand_uniform_double_hq(unsigned int x, unsigned int y)
{
    unsigned long long z = (unsigned long long)x ^
        ((unsigned long long)y << (53 - 32));
    return z * CURAND_2POW53_INV_DOUBLE + (CURAND_2POW53_INV_DOUBLE/2.0);
}
```

**Comment by julian-urban** [comment]
I have zero authority on this subject, but I believe this issue should be closed because it is misleading. After observing different results for a scientific calculation when using numpy vs pytorch uniform sampling at extremely high statistics, finding this open issue led us to believe that this is due to the missing precision close to 0. However, as @colesbury has pointed out, in this regard both libraries appear to have identical behavior, and only after digging into this thread a little more I realized that we drew the wrong conclusion. You could argue that maybe I should just learn how to read properly---and you might be right---but this issue (which should be a non-issue IMHO) was responsible for us working under wrong assumptions for quite a while. I also support the viewpoint that consistency across different math libraries is more important than fixing this known problem in just one of them.

Nevertheless, I do think that numpy handles this better because it doesn't even expose the uniform distribution with single precision to the casual user of `np.random.uniform`, whereas pytorch does. And this can cause problems already at what I would consider low levels of statistics (~1e8) as @obilaniu wrote in the original post. But we were using double in both cases anyway, so this doesn't even apply.

**Comment by obilaniu** [comment]
@julian-urban I would find that a bit unfair. To this day, the issue exists in PyTorch, and although it turned out to not be what impacted your use-case, it absolutely did impact mine and anyone else still relying on the quality of `float32` uniforms who may be impacted in more subtle ways than a NaN. And it could yet be the cause of future problems, so this issue can serve as a sign-post.

I would have filed the same issue under Numpy had I been using Numpy in the counter-factual world where it too was using `float32` to generate `float32` uniforms in its RNG. But as you point out, Numpy only exposes `float64` for that purpose. That increases the threshold for observability from about 2^24 to 2^53 samples, and this makes a very real difference in practice because in a numerical experiment, it is extremely common to sample (far) more than 16 million uniforms, yet (far) less than 8 quadrillion.

**Comment by julian-urban** [comment]
@obilaniu I understand your viewpoint and generally agree with what you're saying. I just think that having an open github issue is not a good way of bringing potentially-affected-people's attention to this problem, as this thread will mostly be discovered after the fact, when an issue has already been detected in a calculation. I would also personally prefer it if pytorch followed the 'numpy way' of hiding the `float32` uniform distribution, but this may be a bottleneck for users who just don't care about the affected precision. Perhaps it would make sense to print a warning when `torch.rand` or `uniform_` are called with `float32` for the first time in a program, or something along those lines. I'm not sure what the best solution is and this is beyond my limited expertise.

I just don't think this issue should remain open indefinitely, as this is a known problem that affects many libraries and experts like you and others in this thread are certainly aware of it. But I get where you're coming from and your viewpoint is completely valid. Maybe others can weigh in with their opinions.

**Comment by obilaniu** [comment]
If we want to avoid keeping this bug permanently open, I'd prefer the solution I proposed (clever use of rounding modes to select with low probability many more FP32 values) or something equivalent;

But if that is too complex to implement, I'd be content with internally using FP64 and casting down into FP32, thus imitating inside PyTorch what Numpy pushes outside itself and making the issue far less salient. It's too late to go back on PyTorch's `uniform_()` for `float32`.

**Comment by colesbury** [comment]
@obilaniu, can you be explicit about your use case? The original issue mentions `log(uniform())` -- that samples from the exponential distribution (and is already provided by [`Tensor.exponential_`](https://pytorch.org/docs/stable/generated/torch.Tensor.exponential_.html)). Was there another use case?

**Comment by obilaniu** [comment]
@colesbury At the time, 2018 and 2019, I was involved in a DL research project on causality over DAGs of categorical random variables. We were using the Gumbel Softmax Trick for several reasons: Partly because it was easy, partly because the neural networks involved produced logits on output, partly because we wanted differentiability, and partly because we needed the "temperature" knob that it provides.

Our algorithm required *extremely* heavy-duty sampling of good-quality, high-entropy Gumbel variates to sample at least some "rare" simulated events for that purpose. Because the transform formula for a standard Gumbel variate is `-log(-log(Uniform(0,1)))`, it is very "squashing" (many uniforms map to the same value) and very sensitive to generation of 0 uniforms, which should have been almost impossible to generate with good-quality RNG.

At the time I wasn't using `.exponential_()`. That might have partially solved our problem, assuming it was good quality itself, but because of that second squashing `-log()` it wouldn't have solved it entirely.

In the end I compensated for this issue mostly by hand-coding in C x86 instrinsics an AVX2+FMA kernel that implemented `-log(-log(u))` as directly as possible, sampling four 32-bit words of entropy from `xorshift128+` and using those for manual generation of `float32` Gumbels.

I still think that the 24 bits of entropy in PyTorch `.uniform_()` are atrocious and given that PRNG is almost memory-bandwidth-limited, drawing more entropy for every float will hardly be a bottleneck in practice.

**Comment by colesbury** [comment]
@obilaniu Thanks, that's helpful. I see how this undersamples low-probability events, but the proposed solution (sampling more densely near 0 for Uniform()) is not going to help. You can verify this empirically for the [binomial distribution](https://gist.github.com/colesbury/b8540be6a2535520abd02f5cf5bff7e0). The density near 0.0 doesn't improve the sampling of low probability events.

A few other minor points:
1) Given the current implementation of uniform, the transform maps unique values to unique values (it's not squashing). See [here](https://colab.research.google.com/drive/1XgS-5J5RCM7OK2UCs2Pryab-Qrgv5YAk?usp=sharing). (It would "squash" values if we sampled more densely near zero, but that just seems like additional evidence for why extra density near zero is not useful.)
2) It's reasonable to think that 24 bits of entropy is atrocious, but then don't use float32. Sampling the uniform more densely near zero doesn't help much -- you only get up to [25 bits of entropy](https://gist.github.com/colesbury/96650ffd8737b048671352ed8bb4d97a)! 

I still have not seen a use case where extra density near zero matters.
